### PR TITLE
Move to native promises / promise-based smartstore and force clients

### PR DESCRIPTION
**Pull request by wmathurin**
* smartsync.js no longer depends on jquery - instead it uses native promises 
* tests no longer use jquery
* userlist sample app no longer uses jquery however sample apps that use Backbone.View still need jquery

* there is now a smartstoreClient (navigator.smartstoreClient or cordova.require("com.salesforce.plugin.smartstore.client") which offers promise-based versions of smartstore methods. (Previously we had such an object in smartsync.js, and we were using it in tests).
* there is now a forceJsClient (defined in force.js) which offers promise-based versions of force's async methods. (Previously we had such an object in smartsync.js, and we were using it in tests).

* NB: Native promises are supported on iOS and on Android 20 (4.4.4) and above.
Currently, our min Android version is 19 (4.4) - so we should either move it up to 20 or document the need for the promise polyfill on Android 19.

### CODE REVIEW

File: gen/plugins/com.salesforce/com.salesforce.plugin.smartstore.js [hunk 1]
```diff
@@ -389,7 +389,7 @@ var moveCursorToNextPage = function (isGlobalStore, cursor, successCB, errorCB)
     if (checkFirstArg(arguments)) return;
     var newPageIndex = cursor.currentPageIndex + 1;
     if (newPageIndex >= cursor.totalPages) {
-        errorCB(cursor, new Error("moveCursorToNextPage called while on last page"));
```
Comment (wmathurin) [inline gen/plugins/com.salesforce/com.salesforce.plugin.smartstore.js]: To be adpatable into a promise, the error callback needs to return only one object (the reason of the failure).

File: gen/plugins/com.salesforce/com.salesforce.util.promiser.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+ * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+// Version this js was shipped with
+var SALESFORCE_MOBILE_SDK_VERSION = "5.0.0";
+
+var promiser = function(object, methodName, objectName) {
```
Comment (wmathurin) [inline gen/plugins/com.salesforce/com.salesforce.util.promiser.js]: That function was previously a private function in smartsync.js

File: test/forcepluginstest/index.html [hunk 3]
```diff
@@ -5,7 +5,6 @@
     <meta http-equiv="Content-type" content="text/html; charset=utf-8">
 
     <!-- external libs -->
-    <script src="dependencies/jquery/jquery.min.js"></script>
```
Comment (wmathurin) [inline test/forcepluginstest/index.html]: No jquery needed to run tests

**Comment by khawkins** [comment]
This looks good to me.  Nice to get rid of jQuery and move to native Promises!

**Pull request merged**